lib/GenomeFileUtil/core/GenomeInterface.py

```diff
@@ -128,54 +171,70 @@ def get_one_genome(self, params):
-        self._own_handle(data, 'genbank_handle_ref')
-        self._own_handle(data, 'gff_handle_ref')
-        if "AnnotatedMetagenomeAssembly" not in ws_datatype:
-            self._check_dna_sequence_in_features(data)
-            data['warnings'] = self.validate_genome(data)
-
-        # sort data
-        data = GenomeUtils.sort_dict(data)
-        # dump genome to scratch for upload
-        data_path = os.path.join(self.scratch, name + ".json")
-        json.dump(data, open(data_path, 'w'))
-        if 'hidden' in params and str(params['hidden']).lower() in ('yes', 'true', 't', '1'):
-            hidden = 1
-        else:
-            hidden = 0
+    def _save_genome_mass(self, params, validate_genome=True):
 
-        if isinstance(workspace, int) or workspace.isdigit():
-            workspace_id = workspace
-        else:
-            workspace_id = self.dfu.ws_name_to_id(workspace)
-
-        save_params = {'id': workspace_id,
-                       'objects': [{'type': ws_datatype,
-                                    'data_json_file': data_path,
-                                    'name': name,
-                                    'meta': meta,
-                                    'hidden': hidden}]}
-        dfu_oi = self.ws_large_data.save_objects(save_params)[0]
-        returnVal = {'info': dfu_oi, 'warnings': data.get('warnings', [])}
-        return returnVal
+        workspace_id = params[_WSID]
+        inputs = params[_INPUTS]
+
+        objects = []
+        warnings = []
+
+        for input_params in inputs:
+
+            obj = {}
+
+            # retrive required params
+            name = input_params['name']
+            data = dict(input_params['data'])
+
+            # XXX there is no `workspace_datatype` param in the spec
+            # NOTE: This allows a user to specify any arbitrary workspace type which could cause,
+            # in the worst case, data corruption. It should be removed from the API
+            # (note it is not currently documented) so users cannot access it.
+            ws_datatype = input_params.get('workspace_datatype', "KBaseGenomes.Genome")
+            # XXX there is no `meta` param in the spec
+            meta = input_params.get('meta', {})
+
+            obj["type"] = ws_datatype
+            obj["name"] = name
+            obj["meta"] = meta
+
+            if "AnnotatedMetagenomeAssembly" not in ws_datatype:
+                if input_params.get('upgrade') or 'feature_counts' not in data:
+                    data = self._update_genome(data)
+
+            # check all handles point to shock nodes owned by calling user
+            self._own_handle(data, 'genbank_handle_ref')
+            self._own_handle(data, 'gff_handle_ref')
+            if "AnnotatedMetagenomeAssembly" not in ws_datatype and validate_genome:
+                self.check_dna_sequence_in_features(data)
+                data['warnings'] = self.validate_genome(data)
```

Comment:
Right, we're skipping these because they could fail, which is why they were moved earlier in the flow, prior to saving assemblies in the blobstore